### PR TITLE
[Feature] Add deposit to fast-unstake

### PR DESCRIPTION
**Pull request by ruseinov**

### CODE REVIEW

File: frame/fast-unstake/src/lib.rs [hunk 1]
```diff
@@ -113,10 +113,12 @@ pub mod pallet {
 			+ IsType<<Self as frame_system::Config>::RuntimeEvent>
 			+ TryInto<Event<Self>>;
 
-		/// The amount of balance slashed per each era that was wastefully checked.
-		///
-		/// A reasonable value could be `runtime_weight_to_fee(weight_per_era_check)`.
-		type SlashPerEra: Get<BalanceOf<Self>>;
+		/// The currency used for deposits.
+		type DepositCurrency: ReservableCurrency<Self::AccountId, Balance = Self::CurrencyBalance>;
```
Comment (kianenigma) [inline frame/fast-unstake/src/lib.rs:117]: I don't see `CurrencyBalance` defined here.

Comment (kianenigma) [inline frame/fast-unstake/src/lib.rs:117]:
```suggestion
		type Currency: ReservableCurrency<Self::AccountId, Balance = Self::CurrencyBalance>;
```

File: frame/fast-unstake/src/lib.rs [hunk 2]
```diff
@@ -381,9 +390,12 @@ pub mod pallet {
 					num_slashing_spans,
 				);
 
+				T::DepositCurrency::unreserve(&stash, deposit.into());
```
Comment (kianenigma) [inline frame/fast-unstake/src/lib.rs:399]: please add `defensive` checks on the return value of this, making sure the remainder is zero.

File: frame/staking/src/pallet/mod.rs [hunk 3]
```diff
@@ -90,6 +90,7 @@ pub mod pallet {
 			+ sp_std::fmt::Debug
 			+ Default
 			+ From<u64>
+			+ Into<u64>
```
Comment (kianenigma) [inline frame/staking/src/pallet/mod.rs:93]: why is this needed?

File: frame/fast-unstake/src/lib.rs [hunk 4]
```diff
@@ -214,6 +216,8 @@ pub mod pallet {
 		/// the chain's resources.
 		#[pallet::weight(<T as Config>::WeightInfo::register_fast_unstake())]
 		pub fn register_fast_unstake(origin: OriginFor<T>) -> DispatchResult {
+			ensure!(ErasToCheckPerBlock::<T>::get() != 0, <Error<T>>::CallNotAllowed);
```
Comment (kianenigma) [inline frame/fast-unstake/src/lib.rs:219]: should be after origin check -- origin check does not need storage read and MUST always be the first check. Otherwise this is DOS possibility: someone sending continuous bar transactions with e.g. origin `None` and causing useless storage reads on nodes.

File: frame/fast-unstake/src/lib.rs (hunk 1 above)
Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:116]: This is already defined in pallet_staking which inherited by this one.

File: frame/fast-unstake/src/lib.rs (hunk 4 above)
Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:218]: Yes, so I had this idea too, but `just a ensure!(ErasToCheckPerBlock != 0, CallNotAllowed) at the beginning of both calls` confused me.

File: frame/fast-unstake/src/lib.rs [hunk 5]
```diff
@@ -249,6 +253,8 @@ pub mod pallet {
 		/// `Staking::rebond`.
 		#[pallet::weight(<T as Config>::WeightInfo::deregister())]
 		pub fn deregister(origin: OriginFor<T>) -> DispatchResult {
+			ensure!(ErasToCheckPerBlock::<T>::get() != 0, <Error<T>>::CallNotAllowed);
```
[the remark this answers is not in the document]
Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:255]: See the comment above, will fix.

File: frame/staking/src/pallet/mod.rs (hunk 3 above)
Comment (ruseinov) [reply frame/staking/src/pallet/mod.rs:93]: Just for convenience to pass the deposit from UnstakeRequest `deposit.into()` to balance operations.

Comment (ruseinov) [reply frame/staking/src/pallet/mod.rs:93]: I'm sure there must be a better solution to this

File: frame/fast-unstake/src/lib.rs (hunk 1 above)
Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:116]: It's inherited from `pallet_staking`

Comment (kianenigma) [reply frame/fast-unstake/src/lib.rs:116]: I'd use `BalanceOf<Self>` to avoid confusion, `BalanceOf<Self>` is statically the same as `Self::CurrencyBalance`.

Comment (kianenigma) [reply frame/fast-unstake/src/lib.rs:116]: You can disambiguate by stating `<T as Config>::Currency` instead of `T::Currency`. Do whichever makes more sense to you, no strong opinion.

File: frame/staking/src/pallet/mod.rs (hunk 3 above)
Comment (ruseinov) [reply frame/staking/src/pallet/mod.rs:93]: The initial reason was the fact that passing `T: pallet_staking::Config` to `UnstakeRequest` results in a `TypeInfo` generation error.

File: frame/fast-unstake/src/lib.rs [hunk 6]
```diff
@@ -381,9 +396,16 @@ pub mod pallet {
 					num_slashing_spans,
 				);
 
-				log!(info, "unstaked {:?}, outcome: {:?}", stash, result);
+				let unreserved = T::DepositCurrency::unreserve(&stash, deposit);
+				if !unreserved.is_zero() {
```
Comment (kianenigma) [inline frame/fast-unstake/src/lib.rs:400]:
```suggestion
				let remaining = T::DepositCurrency::unreserve(&stash, deposit);
				if !remaining.is_zero() {
```

File: frame/fast-unstake/src/mock.rs [hunk 7]
```diff
@@ -164,12 +164,13 @@ impl Convert<sp_core::U256, Balance> for U256ToBalance {
 }
 
 parameter_types! {
-	pub static SlashPerEra: u32 = 100;
+	pub static DepositAmount: u32 = 28_000;
```
Comment (kianenigma) [inline frame/fast-unstake/src/mock.rs:167]: why are we doing this? please keep the test accounts as simple as possible. Sounds like a irrelevant nit, but I have actually learned the hard way that it becomes a PITA very soon.

File: frame/fast-unstake/src/tests.rs [hunk 8]
```diff
@@ -35,16 +35,28 @@ fn test_setup_works() {
...
+		ErasToCheckPerBlock::<T>::put(1);
 		// Controller account registers for fast unstake.
 		assert_ok!(FastUnstake::register_fast_unstake(RuntimeOrigin::signed(2)));
 		// Ensure stash is in the queue.
 		assert_ne!(Queue::<T>::get(1), None);
 	});
 }
 
+#[test]
+fn register_unsigned_fails() {
```
Comment (kianenigma) [inline frame/fast-unstake/src/tests.rs:47]: This test is not doing what the name says -- expecting unsigned origin? or rename the test.

File: frame/fast-unstake/src/tests.rs [hunk 9]
```diff
@@ -60,8 +72,9 @@ fn cannot_register_if_not_bonded() {
 #[test]
 fn cannot_register_if_in_queue() {
 	ExtBuilder::default().build_and_execute(|| {
+		ErasToCheckPerBlock::<T>::put(1);
 		// Insert some Queue item
-		Queue::<T>::insert(1, ());
+		Queue::<T>::insert(1, 10);
```
Comment (kianenigma) [inline frame/fast-unstake/src/tests.rs:77]:
```suggestion
		Queue::<T>::insert(1, Deposit::get());
```

File: frame/fast-unstake/src/tests.rs [hunk 10]
```diff
@@ -129,10 +154,11 @@ fn cannot_deregister_if_not_queued() {
 #[test]
 fn cannot_deregister_already_head() {
 	ExtBuilder::default().build_and_execute(|| {
+		ErasToCheckPerBlock::<T>::put(1);
 		// Controller attempts to register, should fail
 		assert_ok!(FastUnstake::register_fast_unstake(RuntimeOrigin::signed(2)));
 		// Insert some Head item for stash.
-		Head::<T>::put(UnstakeRequest { stash: 1, checked: bounded_vec![] });
+		Head::<T>::put(UnstakeRequest { stash: 1, checked: bounded_vec![], deposit: 28_000 });
```
Comment (kianenigma) [inline frame/fast-unstake/src/tests.rs:161]:
```suggestion
		Head::<T>::put(UnstakeRequest { stash: 1, checked: bounded_vec![], deposit: Deposit::get() });
```

File: frame/fast-unstake/src/tests.rs [hunk 11]
```diff
@@ -165,14 +191,14 @@ mod on_idle {
 
 			// set up Queue item
 			assert_ok!(FastUnstake::register_fast_unstake(RuntimeOrigin::signed(2)));
-			assert_eq!(Queue::<T>::get(1), Some(()));
+			assert_eq!(Queue::<T>::get(1), Some(28_000));
```
Comment (kianenigma) [inline frame/fast-unstake/src/tests.rs:194]:
```suggestion
			assert_eq!(Queue::<T>::get(1), Some(Deposit::get()));
```

None of this should be hardcoded.

**Review by kianenigma — CHANGES_REQUESTED** [review]
Logic looks good, but please:

1. use a smaller number for deposit in mock.rs
2. don't hardcode the value anywhere (unless if really needed)

File: frame/fast-unstake/src/tests.rs [hunk 12]
```diff
@@ -108,9 +146,17 @@ fn deregister_works() {
 	});
 }
 
+#[test]
+fn deregister_eras_0_fails() {
```
Comment (kianenigma) [inline frame/fast-unstake/src/tests.rs:150]:
```suggestion
fn deregister_fails_if_era_to_check_not_set() {
```

and the correct(er) way to test this is to register someone successfully, and see that they cannot even be deregistered if the pallet is in halt.

Comment (kianenigma) [reply frame/fast-unstake/src/tests.rs:150]: well, maybe we can allow deregister at all times and only block register if the pallet is halted?

File: frame/fast-unstake/src/lib.rs [hunk 13]
```diff
@@ -254,7 +264,17 @@ pub mod pallet {
...
+			let deposit = Queue::<T>::take(stash.clone());
+
+			if let Some(deposit) = deposit {
+				let remaining = T::DepositCurrency::unreserve(&stash, deposit);
+				if !remaining.is_zero() {
+					frame_support::defensive!("`not enough balance to unreserve`");
+					ErasToCheckPerBlock::<T>::put(0);
+					Self::deposit_event(Event::<T>::InternalError)
+				}
+			}
```
Comment (kianenigma) [inline frame/fast-unstake/src/lib.rs:276]:
```suggestion
			if let Some(deposit) = deposit.defensive() {
				let remaining = T::DepositCurrency::unreserve(&stash, deposit);
				if !remaining.is_zero() {
					frame_support::defensive!("`not enough balance to unreserve`");
					ErasToCheckPerBlock::<T>::put(0);
					Self::deposit_event(Event::<T>::InternalError)
				}
			}
```

File: frame/fast-unstake/src/tests.rs [hunk 14]
```diff
@@ -35,16 +37,45 @@ fn test_setup_works() {
...
+			FastUnstake::register_fast_unstake(RuntimeOrigin::signed(2)),
+			BalancesError::<T, _>::InsufficientBalance,
+		);
+		// Ensure stash is in the queue.
+		assert_ne!(Queue::<T>::get(1), None);
+	});
+}
+
+#[test]
+fn register_eras_0_fails() {
```
Comment (kianenigma) [inline frame/fast-unstake/src/tests.rs:67]: same as below, not a clear name.

File: frame/fast-unstake/src/tests.rs [hunk 15]
```diff
@@ -35,16 +37,45 @@ fn test_setup_works() {
...
+	use pallet_balances::Error as BalancesError;
+	ExtBuilder::default().build_and_execute(|| {
+		ErasToCheckPerBlock::<T>::put(1);
+		<T as Config>::DepositCurrency::make_free_balance_be(&1, 0);
+
+		// Controller account registers for fast unstake.
+		assert_err!(
+			FastUnstake::register_fast_unstake(RuntimeOrigin::signed(2)),
+			BalancesError::<T, _>::InsufficientBalance,
+		);
```
Comment (kianenigma) [inline frame/fast-unstake/src/tests.rs:59]:
```suggestion
		assert_noop!(
			FastUnstake::register_fast_unstake(RuntimeOrigin::signed(2)),
			BalancesError::<T, _>::InsufficientBalance,
		);
```

**Review by kianenigma — APPROVED** [review]

File: frame/fast-unstake/src/lib.rs [hunk 16]
```diff
@@ -177,6 +181,8 @@ pub mod pallet {
 		NotQueued,
 		/// The provided un-staker is already in Head, and cannot deregister.
 		AlreadyHead,
+		/// The call is not allowed at this point.
+		CallNotAllowed,
```
Comment (gilescope) [inline frame/fast-unstake/src/lib.rs:185]: People are going to ask why call not allowed. Can we have CallNotAllowedBecauseXyz or something more specific. The more specific error enum variants the better - means less support.

Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:185]: In this case `CallNotAllowed` because the pallet is disabled (via setting eras to check to 0). 
I don't think we can be any more specific in this case tbh, this can happen due to Internal errors or just governance changes.
It also seems to be consistent with other pallets, but indeed not very informative.

File: frame/fast-unstake/src/lib.rs [hunk 17]
```diff
@@ -254,7 +267,17 @@ pub mod pallet {
...
 			);
-			Queue::<T>::remove(stash);
+			let deposit = Queue::<T>::take(stash.clone());
+
+			if let Some(deposit) = deposit.defensive() {
+				let remaining = T::DepositCurrency::unreserve(&stash, deposit);
+				if !remaining.is_zero() {
+					frame_support::defensive!("`not enough balance to unreserve`");
+					ErasToCheckPerBlock::<T>::put(0);
+					Self::deposit_event(Event::<T>::InternalError)
```
Comment (gilescope) [inline frame/fast-unstake/src/lib.rs:277]: This then goes on to be Ok(()) - should it be an error?

Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:277]: That's alright, it has defensive checks, what this basically means is that this should always be a defensive path and if it happens - that'd panic.

Comment (ruseinov) [reply frame/fast-unstake/src/lib.rs:277]: And in prod scenario this simply disables the pallet until further notice.

**Review by gilescope — APPROVED** [review]

File: frame/fast-unstake/src/lib.rs (hunk 17 above)
Comment (kianenigma) [reply frame/fast-unstake/src/lib.rs:277]: yes -- a defensive code path is one that should never happen. We're just coding what should happen in case there's a bug. Returning Ok(()), and 

```
ErasToCheckPerBlock::<T>::put(0);
Self::deposit_event(Event::<T>::InternalError)
```

is as good as it gets.

File: frame/fast-unstake/src/lib.rs (hunk 16 above)
Comment (kianenigma) [reply frame/fast-unstake/src/lib.rs:185]: reasonable to rename I presume.

Comment (kianenigma) [reply frame/fast-unstake/src/lib.rs:185]: The docs are shown in the UIs as well, so we only need to change the doc.

File: frame/fast-unstake/src/lib.rs [hunk 18]
```diff
@@ -177,6 +181,8 @@ pub mod pallet {
 		NotQueued,
 		/// The provided un-staker is already in Head, and cannot deregister.
 		AlreadyHead,
+		/// The call is not allowed at this point.
```
Comment (kianenigma) [inline frame/fast-unstake/src/lib.rs:184]:
```suggestion
		/// The call is not allowed at this point because the pallet is not active. 
```

**Comment by ruseinov** [comment]
bot merge
